File: vital-configs/src/main/java/me/vitalframework/configs/processor/VitalPropertiesConfigFileProcessor.java
package me.vitalframework.configs.processor;

import lombok.Getter;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import me.vitalframework.configs.VitalConfig;

import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.util.*;

@RequiredArgsConstructor
@Getter
public class VitalPropertiesConfigFileProcessor implements VitalConfig.FileProcessor {
    @NonNull
    private final File file;

    @NonNull
    private final Properties properties = new Properties();

    @Override
    public Map<String, String> load(@NonNull Class<?> type) throws Exception {
        properties.load(new FileReader(file));

        return Map.ofEntries(properties.entrySet().toArray(Map.Entry[]::new));
    }

    @Override
    public String read(@NonNull String key) {
        return properties.getProperty(key);
    }

    @Override
    public Object read(@NonNull String key, @NonNull Object def) {
        return properties.getProperty(key, String.valueOf(def));
    }

    @Override
    public void write(@NonNull Map<String, ?> serializedContentMap) {
        serializedContentMap.forEach(this::write);
    }

    @Override
    public void write(@NonNull Object object) {
        write(serialize(object));
    }

    @Override
    public void write(@NonNull String key, @NonNull Object value) {
        properties.setProperty(key, String.valueOf(value));
    }

    @Override
    public void save(@NonNull Map<String, ?> serializedContentMap) throws Exception {
        serializedContentMap.entrySet().stream()
                .map((entry) -> Map.entry(entry.getKey(), String.valueOf(entry.getValue())))
                .forEach((entry) -> properties.setProperty(entry.getKey(), entry.getValue()));

        properties.store(new FileWriter(file), null);
    }

    @Override
    public Map<String, String> serialize(@NonNull Object object) {
        final var stringObjectMap = new HashMap<String, String>();

        getPropertyFieldsFromType(object.getClass()).stream()
                .filter(field -> String.class.isAssignableFrom(field.getType()))
                .map(field -> {
                    try {
                        // else use default snakeyaml mapping.
                        return new AbstractMap.SimpleEntry<>(field.getName(), field.get(object));
                    } catch (IllegalAccessException e) {
                        e.printStackTrace();
                        throw new RuntimeException("error while serializing properties config field %s"
                                .formatted(field.getName()));
                    }
                })
                .forEach((entry) -> stringObjectMap.put(entry.getKey(), (String) entry.getValue()));

        return stringObjectMap;
    }

    @Override
    public Object deserialize(@NonNull Map<String, ?> serializedContentMap, @NonNull Class<Object> type) throws Exception {
        try {
            final var defaultConstructor = type.getConstructor();
            final var object = defaultConstructor.newInstance();

            // default constructor was found, inject field properties...
            serializedContentMap
                    .forEach((key, value) -> {
                        final var optionalField = Optional.ofNullable(getFieldByProperty(type, key));

                        optionalField.ifPresent(field -> VitalConfig.injectField(object, field, value));
                    });

            return object;
        } catch (NoSuchMethodException e) {
            // default constructor not found, attempt to get constructor matching properties...
            final var constructor = type.getConstructor(getPropertyFieldsFromType(type).stream().map(Object::getClass).toArray(Class[]::new));

            // constructor found, create new instance with this constructor...
            return constructor.newInstance(serializedContentMap.values());
        }
    }
}

File: vital-configs/src/main/kotlin/me/vitalframework/configs/processor/VitalPropertiesConfigFileProcessor.kt
package me.vitalframework.configs.processor

import me.vitalframework.configs.VitalConfig.Companion.injectField
import me.vitalframework.configs.VitalConfig.FileProcessor
import java.io.File
import java.io.FileReader
import java.io.FileWriter
import java.util.*

class VitalPropertiesConfigFileProcessor(
    override val file: File,
) : FileProcessor<String> {
    val properties = Properties()

    @Throws(Exception::class)
    override fun load(type: Class<*>): Map<String, String> {
        properties.load(FileReader(file))

        return mapOf(*properties.entries.map { (key, value) -> key.toString() to value.toString() }.toTypedArray())
    }

    override fun read(key: String): String? {
        return properties.getProperty(key)
    }

    override fun read(key: String, def: String): String? {
        return properties.getProperty(key, def.toString())
    }

    override fun write(serializedContent: Map<String, String>) {
        serializedContent.forEach { (key, value) ->
            this.write(key, value)
        }
    }

    override fun write(instance: Any) {
        write(serialize(instance))
    }

    override fun write(key: String, value: String) {
        properties.setProperty(key, value)
    }

    @Throws(Exception::class)
    override fun save(serializedContent: Map<String, String>) {
        serializedContent.entries.forEach { (key, value) ->
            properties.setProperty(key, value)
        }

        properties.store(FileWriter(file), null)
    }

    override fun serialize(instance: Any): Map<String, String> {
        val stringObject = mutableMapOf<String, String>()

        getPropertyFieldsFromType(instance.javaClass)
            .filter { String::class.java.isAssignableFrom(it.type) }
            .map {
                try {
                    // else use default snakeyaml mapping.
                    return@map it.name to it[instance]
                } catch (e: IllegalAccessException) {
                    e.printStackTrace()
                    throw RuntimeException("error while serializing properties config field '${it.name}'")
                }
            }
            .forEach { (key, value) ->
                stringObject[key] = value as String
            }

        return stringObject
    }

    @Throws(Exception::class)
    override fun deserialize(serializedContent: Map<String, String>, type: Class<*>): Any? {
        try {
            val defaultConstructor = type.getConstructor()
            val instance = defaultConstructor.newInstance()

            // default constructor was found, inject field properties...
            serializedContent
                .forEach { (key, value) ->
                    getFieldByProperty(type, key)?.let {
                        injectField(
                            instance,
                            it,
                            value
                        )
                    }
                }

            return instance
        } catch (e: NoSuchMethodException) {
            // default constructor not found, attempt to get constructor matching properties...
            val constructor = type.getConstructor(*getPropertyFieldsFromType(type).map { it.javaClass }.toTypedArray())

            // constructor found, create new instance with this constructor...
            return constructor.newInstance(serializedContent.values)
        }
    }
}